Add OS Secrets RC API Docs (#2144)
hashicorp/web-unified-docs: content/vault/v2.x (rc)/content/api-docs/secret/os.mdx

---
layout: api

> page_title: OS - Secrets Engines - HTTP API

Reviewer: ```suggestion
page_title: OS - Secrets Engines
```
The platform automatically appends breadcrumb info

> description: This is the API documentation for the Vault OS secrets engine.

Reviewer: ```suggestion
description: >-
  Use the OS plugin API to manage user credentials on remote hosts for RHEL
  systems.
```
Style correction: describe the problem readers can solve rather than the doc itself

---

# OS secrets engine (API)

This is the API documentation for the Vault OS secrets engine. For general
information about the usage and operation of the OS secrets engine, please see
the [Vault OS documentation](/vault/docs/secrets/os).

This documentation assumes the OS secrets engine is enabled at the `/os` path
in Vault. Since it is possible to enable secrets engines at any location, please
update your API calls accordingly.

## Configure top level parameters

| Method | Path         |
|:-------|:-------------|
| `POST` | `/os/config` |

This endpoint configures settings that apply to all hosts or accounts.

### Parameters

> - `max_versions` `(int: 10)`: The number of secret versions to keep for each host account. This value applies to all keys.

Reviewer: ```suggestion
- `max_versions` `(int: 10)`: The number of secret versions to keep for each
  host account. The max version limit applies to all keys.
```
Style correction: avoid "this" as a pronoun

- `ssh_host_key_trust_on_first_use` `(bool: false)`: Whether to allow trust-on-first-use ssh key adoption. Defaults to false.
- `password_policy` `(string)`: The name of a password policy. Global password policy for this mount.
Can be overridden by host and account level settings.

### Sample payload

```json
{
  "max_versions": 10,
  "ssh_host_key_trust_on_first_use": true,
  "password_policy": "short"
}
```

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request POST \
    --data @payload.json \

>     https://127.0.0.1:8200/v1/os/config

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/os/config
```

```

## Read top level parameters

| Method | Path         |
|:-------|:-------------|
| `GET` | `/os/config` |

Reports configured top level parameters that apply to all configured hosts.

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

>     http://127.0.0.1:8200/v1/os/config

Reviewer: ```suggestion
    ${VAULT_ADDR}/v1/os/config
```

```

### Sample response

```json
  {
    "max_versions": 10,
    "ssh_host_key_trust_on_first_use": true,
    "password_policy": "short"
  }
```

## Create Host

| Method | Path              |
|:-------|:------------------|
| `POST` | `/os/hosts/:name` |

This endpoint configures the connection information used to communicate with the desired target host.

> Note: As per Vault convention, this endpoint handles both the Create and Update operations.

Reviewer: ```suggestion
Configure or update the connection information used to communicate with the
target host.
```

### Parameters

> - `name` `(string: <required>)`: the unique name for the host.

Reviewer: ```suggestion
### Path parameters

- `name` `(string: <required>)`: Unique name for the target host configuration.

### Request parameters

```

Reviewer: From context, I'm assuming this is a name use by Vault to identify the configuration not the actual host (which is identified by the address field below)?

- `address` `(string: <required>)`: hostname or IPv4/IPv6 address of the remote host.

> - `port` `(int: 22)`: Port number for SSH connection. The default is '22'.

Reviewer: ```suggestion
- `port` `(int: 22)`: Port number for SSH connection.
```

### Sample payload

```json
{
  "address": "192.0.2.0",
  "port": 22,
  "ssh_host_key": "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOMqqnkVzrm0SdG6UOoqKLsabgH5C9okWi0dh2l9GKJl",
  "password_policy": "short",
  "custom_metadata": {
    "foo": "bar",
    "baz": 67
  }
}
```

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request POST \
    --data @payload.json \
    https://127.0.0.1:8200/v1/os/hosts/jumpbox
```

## Read Host

| Method | Path              |
|:-------|:------------------|
| `GET`  | `/os/hosts/:name` |

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    http://127.0.0.1:8200/v1/os/hosts/jumpbox
```

### Sample response

```json
{
  "address": "192.0.2.0",
  "port": 22,
  "ssh_host_key": "ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOMqqnkVzrm0SdG6UOoqKLsabgH5C9okWi0dh2l9GKJl",
  "password_policy": "short",
  "custom_metadata": {
    "foo": "bar",
    "baz": 67
  }
}
```

## Delete Host

| Method | Path              |
|:-------|:------------------|
| `DELETE`  | `/os/hosts/:name` |

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request DELETE
    http://127.0.0.1:8200/v1/os/hosts/jumpbox
```

## List Hosts

| Method | Path              |
|:-------|:------------------|
| `LIST`  | `/os/hosts` |

> 

Author: added

### Sample request

```shell-session
$ curl \

>     --header "X-Vault-Token: ..." \

Reviewer: ```suggestion
    --header "X-Vault-Token: ${VAULT_TOKEN}" \
```

    --request LIST
    http://127.0.0.1:8200/v1/os/hosts
```

### Sample response

```json
{
  "data": {
    "keys": ["jumpbox"]
  }
}
```

## Create Account

| Method | Path              |
|:-------|:------------------|
| `POST` | `/os/hosts/:host_ref/accounts/:name` |

> This endpoint configures an individual account on an already created host.

Reviewer: ```suggestion
Configure an individual account on an existing host.
```

- `password` `(string: <required>)`: The initial password of the local account. Cannot be changed after creation via this path (see /reset).
- `parent_account_ref` `(string)`: Reference to the name of the account for this host used to rotate this account. If not provided,
this account will rotate its own credentials. (See: TBD in docs?)
- `password_policy` `(string)`: Name of a password policy. Will override any policy set in the associated host.
- `custom_metadata` `(map<string|string>)`: A map of arbitrary string to string valued user-provided metadata meant to describe the secret.
Note that on updates, the entire map will be replaced with the new map, if one is provided.
@include 'rotationfields.mdx'

Rotation fields set here will override any rotation settings configured on the associated host.

### Sample payload

```json
{
  "username": "dave",
  "password": "hunter2",
  "password_policy": "short",
  "parent_account_ref": "other_dave",
  "rotation_schedule": "0 0 1 * *",
  "custom_metadata": {
    "foo": "bar",
    "baz": 67
  }
}
```

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request POST \
    --data @payload.json \

>     https://127.0.0.1:8200/v1/os/hosts/jumpbox/account/dave

Reviewer 2: ```suggestion
    https://127.0.0.1:8200/v1/os/hosts/jumpbox/accounts/dave
```

```

## Read Account

| Method | Path              |
|:-------|:------------------|
| `GET`  | `/os/hosts/:host_ref/accounts/:name` |

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    http://127.0.0.1:8200/v1/os/hosts/jumpbox/accounts/dave
```

### Sample response

```json
{
  "username": "dave",
  "last_vault_rotation": "2026-01-01T00:00:00Z",
  "next_vault_rotation": "2026-02-01T00:00:00Z",
  "parent_account_ref": "other_dave",
  "rotation_schedule": "0 0 1 * *",
  "password_policy": "short",
  "custom_metadata": {
    "foo": "bar",
    "baz": 67
  }

> }

Reviewer 2: ```suggestion
{
  "request_id": "754171a3-9557-f4b2-1322-d9d8db4ed3da",
  "lease_id": "",
  "renewable": false,
  "lease_duration": 0,
  "data": {
    "current_version": 1,
    "custom_metadata": {
        "foo": "bar",
        "baz": 67
    },
    "disable_automated_rotation": false,
    "last_vault_rotation": "2026-01-01T00:00:00Z",
    "next_vault_rotation": "2026-02-01T00:00:00Z",
    "parent_account_ref": "other_dave",
    "password_policy": "short",
    "rotation_period": 0,
    "rotation_policy": "",
    "rotation_schedule": "0 0 1 * *",
    "rotation_window": 0,
    "ttl": 0,
    "username": "dave"
  },
  "wrap_info": null,
  "warnings": null,
  "auth": null,
  "mount_type": "vault-plugin-secrets-os"
}

```

```

## Delete Account

| Method | Path              |
|:-------|:------------------|
| `DELETE`  | `/os/hosts/:host_ref/accounts/:name` |

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request DELETE
    http://127.0.0.1:8200/v1/os/hosts/jumpbox/accounts/dave
```

## List Accounts

| Method | Path              |
|:-------|:------------------|
| `LIST`  | `/os/hosts/:host_ref/accounts/` |

### Sample request

```shell-session
$ curl \
    --header "X-Vault-Token: ..." \
    --request LIST
    http://127.0.0.1:8200/v1/os/hosts/jumpbox/accounts/
```

### Sample response

```json
{
  "data": {
    "keys": ["dave", "other_dave"]
  }

> }

Author: @schavis any thoughts on using elided responses or full responses here? (Or updating the example shell command to use `jq .data`)?
